runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java

```diff
@@ -90,4 +94,42 @@ public void testUtf8UserNameAndApplicationNameIsNormalized() {
     options.setAppName("fəˈnɛtık əsoʊsiˈeıʃn");
     assertEquals("f00n0t0k00so0si0e00n-0i00nt00n000n0l0-1208190706", options.getJobName());
   }
+
+  @Test
+  public void testStagingLocation() {
+    DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+    IOChannelUtils.registerStandardIOFactories(options);
+    options.setTempLocation("file://temp_location");
+    options.setStagingLocation("gs://staging_location");
+    assertTrue(isNullOrEmpty(options.getGcpTempLocation()));
+    assertEquals("gs://staging_location", options.getStagingLocation());
+  }
+
+  @Test
+  public void testDefaultStagingLocation() {
+    DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+    IOChannelUtils.registerStandardIOFactories(options);
+    options.setTempLocation("gs://temp_location");
+    assertEquals("gs://temp_location", options.getGcpTempLocation());
+    assertEquals("gs://temp_location/staging", options.getStagingLocation());
+  }
+
+  @Test
+  public void testDefaultNoneGcsTempLocation() {
+    DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+    options.setTempLocation("file://temp_location");
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Missing required value: at least one of tempLocation, gcpTempLocation"
+        + " or stagingLocation must be set.");
+    options.getStagingLocation();
+  }
+
+  @Test
+  public void testDefaultStagingLocationUnset() {
+    DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Missing required value: at least one of tempLocation, gcpTempLocation"
+        + " or stagingLocation must be set.");
+    options.getStagingLocation();
+  }
```

Comment:
test where staging defaults to gcsTemp

e.g., temp is a gs://a, gcsTemp is gs://b, and staging picks gs://b/staging